kmk/kmk_keyboard.py

```diff
@@ -90,223 +87,354 @@ def __repr__(self):
-                self.split_target_left = not self.is_target
+                    self.coord_mapping.append(intify_coordinate(ridx, cidx))
+
+    def _init_hid(self):
+        if self.hid_type == HIDModes.NOOP:
+            self._hid_helper = AbstractHID
+        elif self.hid_type == HIDModes.USB:
+            self._hid_helper = USBHID
+        elif self.hid_type == HIDModes.BLE:
+            self._hid_helper = BLEHID
         else:
-            self.is_target = True
+            self._hid_helper = AbstractHID
+        self._hid_helper = self._hid_helper()
 
-        if self.uart_pin is not None:
-            self.uart = self.init_uart(self.uart_pin)
-
-        if self.rgb_pixel_pin:
-            self.pixels = rgb.RGB(self.rgb_config, self.rgb_pixel_pin)
-            self.rgb_config = None  # No longer needed
-            self.pixels.loopcounter = 0
-        else:
-            self.pixels = None
-
-        if self.led_pin:
-            self.led = led.led(self.led_pin, self.led_config)
-            self.led_config = None  # No longer needed
-        else:
-            self.led = None
-
-        self.matrix = self.matrix_scanner(
+    def _init_matrix(self):
+        self.matrix = MatrixScanner(
             cols=self.col_pins,
             rows=self.row_pins,
             diode_orientation=self.diode_orientation,
             rollover_cols_every_rows=getattr(self, 'rollover_cols_every_rows', None),
         )
 
-        # Compile string leader sequences
-        for k, v in self.leader_dictionary.items():
-            if not isinstance(k, tuple):
-                new_key = tuple(KC[c] for c in k)
-                self.leader_dictionary[new_key] = v
+        return self
 
-        for k, v in self.leader_dictionary.items():
-            if not isinstance(k, tuple):
-                del self.leader_dictionary[k]
+    # Only one GC to allow for extentions to have room.
+    # There are random memory allocations without this
+    gc.collect()
 
-        while True:
-            if self.split_type is not None and self.is_target:
-                update = self._receive_from_initiator()
-                if update is not None:
-                    self._handle_matrix_report(update)
+    def go(self, hid_type=HIDModes.USB, **kwargs):
+        self.hid_type = hid_type
 
-            update = self.matrix.scan_for_changes()
+        self._init_sanity_check()
+        self._init_coord_mapping()
+        self._init_hid()
 
-            if update is not None:
-                if self.is_target:
-                    self._handle_matrix_report(update)
-                else:
-                    # This keyboard is a initiator, and needs to send data to target
-                    self._send_to_target(update)
+        for ext in self.extensions:
+            try:
+                ext.during_bootup(self)
+            except Exception:
+                print('Failed to load extention', ext)
+                import time
+
+                time.sleep(5)
```

Comment:
time should basically always be abstracted by `kmktime`, yeah? `sleep_ms(5000)`

Reply:
This can probably be removed. It's just been an artifact mostly. Up to you on which you prefer.

Reply:
yeah, if we can remove the sleep, awesome, if not, then let's `kmktime` it

Reply:
Error spew is common when an extension is bad which is why I left the sleep. Maybe I'll leave this in under debug.